src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -0,0 +1,389 @@
+                }
+            }
+
+            try
+            {
+                Log.LogMessage($"Uploading {item} to {uploadPath}.");
+                UploadClient uploadClient = new UploadClient(Log);
+                await
+                    uploadClient.UploadBlockBlobAsync(
+                        ct,
+                        feed.AccountName,
+                        feed.AccountKey,
+                        feed.ContainerName,
+                        item,
+                        uploadPath,
+                        leaseId);
+            }
+            finally
+            {
+                clientThrottle.Release();
+            }
+            if (isLeaseRequired) {
+                using (HttpClient client = new HttpClient())
+                {
+                    try
+                    {
+                        // release a lease on blob, retry if unsuccesful
+                        string leaseUrl = AzureHelper.GetBlobRestUrl(feed.AccountName, feed.ContainerName, uploadPath) + "?comp=lease";
+                        Tuple<string, string> headerLeaseAction = new Tuple<string, string>("x-ms-lease-action", "release");
+                        Tuple<string, string> headerLeaseId = new Tuple<string, string>("x-ms-lease-id", leaseId);
+                        List<Tuple<string, string>> additionalHeaders = new List<Tuple<string, string>>() { headerLeaseAction, headerLeaseId };
+                        var getLeaseRequest = AzureHelper.RequestMessage("PUT", leaseUrl, feed.AccountName, feed.AccountKey, additionalHeaders);
+                        using (HttpResponseMessage response = await AzureHelper.RequestWithRetry(Log, client, getLeaseRequest))
+                        {
+                            if (response.IsSuccessStatusCode)
+                            {
+                                Log.LogMessage($"Release lease on {uploadPath}", MessageImportance.Low);
+                            }
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        //Failed after retry 
+                        Log.LogErrorFromException(e);
+                    }
+                }
+            }
+        }
+
+        private async Task MergeIndexes(CancellationToken ct, string item, string relativePath, List<string> indexJsonList)
+        {
+            //pull equivalent index.json
+            Tuple<string, NuGetVersion> blobPath = feed.CalculateBlobPath(item, relativePath);
+            string relativeBlobPath = Path.Combine(relativePath, blobPath.Item1).ToLowerInvariant().Replace("\\", "/");
+            try
+            {
+                if (indexJsonList.Any(x => x.Contains(relativeBlobPath)))
+                {
+                    //download index.json to download folder
+                    string blob = indexJsonList.Find(x => x.Contains(relativeBlobPath));
+                    using (HttpClient client = new HttpClient())
```

Comment:
We need to put this DownloadFromAzure into a helper type and use it instead of duplicating all this code.

Reply:
fixed.